Add tests and support for variables in =~/2

fnogatz opened the pull request:
Fixes #11.

mndrix commented:
Thanks. I should be able to review this on Monday.

>

mndrix commented on t/synopsis.pl:
```diff
@@ -4,3 +4,11 @@
+
+'works with text in variable' :-
+    Text = a,
+    Text =~ '[a-z]'.
```
These tests look good.  The file t/synopsis.pl is for testing the code that's in the [Synopsis section](https://github.com/mndrix/regex/tree/c264f36a5cbbac20c1acafe7b770e7fe93334a7b#synopsis) of the readme file, so these tests should probably be in t/examples.pl or maybe a new file t/regressions.pl

fnogatz replied:
I have moved the tests into `t/examples.pl` now.

mndrix commented on prolog/regex.pl:
```diff
@@ -40,8 +40,8 @@
-    % is goal expansion wanted?
+user:term_expansion(Text =~ Pattern, Goal) :-
+    % is term expansion wanted?
     prolog_load_context(module, Module),
```
When running with this change, I get warnings like the following and 5 tests in t/captures.pl fail:

```
Warning: /home/mndrix/src/pack-regex/t/examples.pl:10:
        Local definition of user:regex/4 overrides weak import from regex
Warning: /home/mndrix/src/pack-regex/t/examples.pl:32:
        Clauses of regex/4 are not together in the source-file
          Earlier definition at /home/mndrix/src/pack-regex/t/examples.pl:10
          Current predicate: 'i beg with ^'/0
          Use :- discontiguous regex/4. to suppress this message
```

In case it helps, I have the following two files in my working directory that I use for testing:

```
$ cat dev.pl
:- assertz(file_search_path(library, '/home/mndrix/src/pack-regex/prolog')).
```

and

```
$ cat Makefile 
test:
        prove -e 'swipl -f dev.pl -q -t main -s' t/*
```

fnogatz commented:
The problem arises as `goal_expansion` only sees a variable on the right-hand side, although it might be already bound in the given context. I have added a `\+var(Pattern)` test and reverted the `term_expansion` change.

However, this leads to another failing test case for captures with a rhs variable. They apparently need the goal expansion for the variable binding, which is not applied when giving a variable on the rhs, due to the other fix.

fnogatz commented:
And, by the way, I am totally fine with the solution to simply use `regex/4` for such use-cases :)

mndrix commented:
Thanks @fnogatz  Your test cases were very helpful.  I've combined those three commits into commit 4099834.  Your partial fix in commit e184922 also lead me to a fix which appears to work in all cases (see commit b2711fb).

If this all looks good to you, I'll close this issue along with #12 and release a new pack with all these changes.

fnogatz commented:
Looks good to me, thank you :)

mndrix commented:
Fixed in release v0.3.3
